ietf/iesg/utils.py

```diff
@@ -39,30 +52,269 @@ def telechat_page_count(date=None, docs=None, ad=None):
         else:
             pass
 
     related_pages = 0
-    for d in for_approval-set(drafts):
-        if d.type_id == 'statchg':
+    for d in for_approval - set(drafts):
+        if d.type_id == "statchg":
             for rel in d.related_that_doc(STATUSCHANGE_RELATIONS):
                 related_pages += rel.pages or 0
-        elif d.type_id == 'conflrev':
-            for rel in d.related_that_doc('conflrev'):
+        elif d.type_id == "conflrev":
+            for rel in d.related_that_doc("conflrev"):
                 related_pages += rel.pages or 0
         else:
             # There's really nothing to rely on to give a reading load estimate for charters
             pass
-    
-    return TelechatPageCount(for_approval=pages_for_approval,
-                             for_action=pages_for_action,
-                             related=related_pages,
-                             ad_pages_left_to_ballot_on=ad_pages_left_to_ballot_on)
+
+    return TelechatPageCount(
+        for_approval=pages_for_approval,
+        for_action=pages_for_action,
+        related=related_pages,
+        ad_pages_left_to_ballot_on=ad_pages_left_to_ballot_on,
+    )
+
+
+def get_wg_dashboard_info():
+    docs = (
+        Document.objects.filter(
+            group__type="wg",
+            group__state="active",
+            states__type="draft",
+            states__slug="active",
+        )
+        .filter(models.Q(ad__isnull=True) | models.Q(ad__in=get_active_ads()))
+        .distinct()
+        .prefetch_related("group", "group__parent")
+        .exclude(
+            states__type="draft-stream-ietf",
+            states__slug__in=["c-adopt", "wg-cand", "dead", "parked", "info"],
+        )
+    )
+    groups = Group.objects.filter(state="active", type="wg")
+    areas = Group.objects.filter(state="active", type="area")
+
+    total_group_count = groups.count()
+    total_doc_count = docs.count()
+    total_page_count = docs.aggregate(models.Sum("pages"))["pages__sum"] or 0
+    totals = {
+        "group_count": total_group_count,
+        "doc_count": total_doc_count,
+        "page_count": total_page_count,
+    }
+
+    # Since this view is primarily about counting subsets of the above docs query and the
+    # expected number of returned documents is just under 1000 typically - do the totaling
+    # work in python rather than asking the db to do it.
+
+    groups_for_area = defaultdict(set)
+    pages_for_area = defaultdict(lambda: 0)
+    docs_for_area = defaultdict(lambda: 0)
+    groups_for_ad = defaultdict(lambda: defaultdict(set))
+    pages_for_ad = defaultdict(lambda: defaultdict(lambda: 0))
+    docs_for_ad = defaultdict(lambda: defaultdict(lambda: 0))
+    groups_for_noad = defaultdict(lambda: defaultdict(set))
+    pages_for_noad = defaultdict(lambda: defaultdict(lambda: 0))
+    docs_for_noad = defaultdict(lambda: defaultdict(lambda: 0))
+    docs_for_wg = defaultdict(lambda: 0)
+    pages_for_wg = defaultdict(lambda: 0)
```

Comment:
defaultdict are a handy but can cause bugs by adding keys if later uses, e.g., have a typo. Once you've built up your data, you can set `groups_for_area.default_factory = None` (etc) and that will effectively turn it into a standard dictionary.  https://docs.python.org/3/library/collections.html#collections.defaultdict.default_factory

Reply:
That's not helpful for the way _these_ dicts are being used. The property that an access for an unknown key returns a 0 is important and is used until the last mention of each of those dictionaries.

Reply:
Agree - I misread the code and thought the defaultdicts were included in the return value.